File: BinaryTree/Leetcode/Medium/SortedListToBST/SortedListToBST.java
// https://leetcode.com/problems/convert-sorted-list-to-binary-search-tree/
class ListNode {

    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }

}

class TreeNode {

    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class SortedListToBST {
    TreeNode convertToBST(ListNode head, ListNode end) {
        if (head == null || head == end) {
            return null;
        }
        ListNode slow = head, fast = head;
        while (fast != end && fast.next != end) {
            slow = slow.next;
            fast = fast.next.next;
        }
        TreeNode root = new TreeNode(slow.val);
        root.left = convertToBST(head, slow);
        root.right = convertToBST(slow.next, end);
        return root;
    }

    public TreeNode sortedListToBST(ListNode head) {
        return convertToBST(head, null);
    }
}

File: BinaryTree/Leetcode/Medium/SortedListToBST/SortedListToBST.cpp
// https://leetcode.com/problems/convert-sorted-list-to-binary-search-tree/
#include <bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
TreeNode *listToBST(ListNode *head, ListNode *end)
{
    if (!head || head == end)
        return nullptr;
    ListNode *slow = head, *fast = head;
    while (fast != end && fast->next != end)
    {
        slow = slow->next;
        fast = fast->next->next;
    }
    TreeNode *root = new TreeNode(slow->val);
    root->left = listToBST(head, slow);
    root->right = listToBST(slow->next, end);
    return root;
}

TreeNode *sortedListToBST(ListNode *head)
{
    return listToBST(head, nullptr);
}